internal/db/schema/migrations/oss/postgres/51/03_create_kms_key_destructions.up.sql

```diff
@@ -0,0 +1,69 @@
+begin;
+
+  create table kms_data_key_version_destruction_job (
+    key_id kms_private_id primary key -- one job per key version
+      references kms_data_key_version (private_id)
+        on delete cascade -- Note: instead of restrict, we cascade the delete to this job when the key version is deleted
+        on update cascade,
+    create_time wt_timestamp not null
+  );
+  comment on table kms_data_key_version_destruction_job is
+    'Table holding running and pending data key version destruction jobs';
+
+  create trigger immutable_columns before update on kms_data_key_version_destruction_job
+    for each row execute procedure immutable_columns('key_id', 'create_time');
+
+  create table kms_data_key_version_destruction_job_run (
+    key_id kms_private_id not null
+      references kms_data_key_version_destruction_job (key_id)
+        on delete cascade
+        on update cascade,
+    table_name text not null,
+    total_count int not null
+      constraint total_count_cannot_be_negative
+        check (total_count > 0), -- Must not be 0
+    completed_count int not null default 0
+      constraint completed_count_cannot_be_negative
+        check (completed_count >= 0),
+    is_running boolean not null default false,
```

Comment:
Does `is_running` ever change from `true` to `false` before `completed_count = total_count`?

Reply:
It's used to guarantee that only one run is running, across all runs in a cluster through the unique index below.

The current schema supports any value of `is_running` on a row as long as `completed_count < total_count`. The application will normally only set `is_running` to `false` if `completed_count=total_count`, but if a run is interrupted mid-execution it could set `is_running` to `false` with `completed_count < total_count`.

Reply:
Does the code detect a run was interrupted and set it back to false?

Reply:
It does detect a run that was interrupted, but it just resumes the run in that case rather than setting it back to false. The logic is here: https://github.com/hashicorp/boundary/blob/64417de79c5f13ffad6accaeaccc2f2168d5038e/internal/kms/kms.go#L434-L485

Reply:
How will the job be resumed if the controller running it fails/exits unexpectedly?

Reply:
It will be resumed by this code the next time the scheduler runs the job for this tablename. The code will detect that it is still supposed to be running and resume the operation.

Reply:
Could the `is_running` column be removed and the value be calculated? For example:
```sql
case when completed_count = total_count then 'completed'
     when completed_count > 0 then 'running'
     else 'pending'
```

Reply:
We currently use `is_running` to assign what run should be running at the _start_ of a job, if we removed it we may have multiple jobs starting at the same time, so I think we will still need it as the DB arbiter of which run gets to start.